src/sched/job_runner.cc

```diff
@@ -196,6 +199,32 @@ void JobRunnerStrand::PushToRunnerIfNeeded(const bool is_in_running_job) {
     }
 }
 
+// Try running job immediately in the same thread.
+//
+// To ensure correctness without performance regression,
+// We only do a quick check to see if it can be run.
+// If not sure, fall back to the normal AddJob.
+JobRunner::TryRunImmediately::State JobRunnerStrand::AddJob(
```

Comment:
- I would say we provide this in the old interface as well, as an execution policy, so that we can have dynamic/runtime control on it.
- Add a deterministic policy, so that we can force local exec. That would help a lot in profiling.

Reply:
> I would say we provide this in the old interface as well, as an execution policy, so that we can have dynamic/runtime control on it.

Will do as a follow-up.

> Add a deterministic policy, so that we can force local exec. That would help a lot in profiling.

What is the point? It may not always be possible to do that. In this case, `FINISHED` means "ok we have finished it locally", "ENQUEUED" means "it may not be possible to run it locally, so fallback to queueing mode". Forcing local exec does not bring anything more.

Reply:
> it may not be possible to run it locally, so fallback to queueing mode

Similarly, "..., so report error instead of falling back."

Reply:
Done